src/ConversationList/ConversationListBox.vala

```diff
@@ -48,26 +51,71 @@ public class Mail.ConversationListBox : Gtk.ListBox {
             cancellable.cancel ();
         }
 
-        get_children ().foreach ((child) => {
-            child.destroy ();
-        });
+        lock (conversations) {
+            conversations.clear ();
+            get_children ().foreach ((child) => {
+                child.destroy ();
+            });
+
+            cancellable = new GLib.Cancellable ();
+            try {
+                folder = yield ((Camel.Store) current_account.service).get_folder (current_folder, 0, GLib.Priority.DEFAULT, cancellable);
+                folder.changed.connect ((change_info) => folder_changed (change_info, cancellable));
+                thread = new Camel.FolderThread (folder, folder.get_uids (), false);
+                unowned Camel.FolderThreadNode? child = (Camel.FolderThreadNode?) thread.tree;
+                while (child != null) {
+                    if (cancellable.is_cancelled ()) {
+                        break;
+                    }
+
+                    add_conversation_item (child);
+                    child = (Camel.FolderThreadNode?) child.next;
+                }
+
+                yield folder.refresh_info (GLib.Priority.DEFAULT, cancellable);
+            } catch (Error e) {
+                critical (e.message);
+            }
+        }
+    }
+
+    private void folder_changed (Camel.FolderChangeInfo change_info, GLib.Cancellable cancellable) {
+        if (cancellable.is_cancelled ())
+            return;
+
+        lock (conversations) {
+            thread.apply (folder.get_uids ());
+            change_info.get_removed_uids ().foreach ((uid) => {
+                var item = conversations[uid];
+                if (item != null) {
+                    conversations.unset (uid);
+                    item.destroy ();
+                }
+            });
 
-        cancellable = new GLib.Cancellable ();
-        try {
-            Camel.Folder? folder = yield ((Camel.Store) current_account.service).get_folder (current_folder, Camel.StoreGetFolderFlags.BODY_INDEX, GLib.Priority.DEFAULT, cancellable);
-            /*yield folder.refresh_info (GLib.Priority.DEFAULT, cancellable);*/
-            thread = new Camel.FolderThread (folder, folder.get_uids (), false);
             unowned Camel.FolderThreadNode? child = (Camel.FolderThreadNode?) thread.tree;
             while (child != null) {
```

Comment:
The problem is that `get_added_uids` doesn't give us in which thread it is added, so we need to iterate the threadnodes again…